Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,261 @@ Primary Objectives
 Secondary Objectives
 * No “swizzling” or lying about struct types – they are always struct types
  - No confusing use of GT_LCL_FLD to refer to the entire struct as a different type
+  
+Normalizing Struct Types
+------------------------
+We would like to facilitate full enregistration of structs with the following properties:
+1. Its fields are infrequently accessed, and
+1. The entire struct fits into a register, and
+2. Its value is used or defined in a register 
+(i.e. as an argument to or return value from calls or intrinsics).
+
+In RyuJIT, the concept of a type is very simplistic (which helps support the high throughput
+of the JIT). Rather than a symbol table to hold the properties of a type, RyuJIT primarily
+deals with types as simple values of an enumeration. When more detailed information is
+required about the structure of a type, we query the type system, across the JIT/EE interface.
+This is generally done only during the importer (translation from MSIL to the RyuJIT IR),
+during struct promotion analysis, and when determining how to pass or return struct values.
+As a result, struct types are generally treated as an opaque type
+(TYP_STRUCT) of unknown size and structure.
+
+In order to treat fully-enregisterable struct types as "first class" types in RyuJIT, we
+ created new types to represent vectors, in order for the JIT to support operations on them:
+* `TYP_SIMD8`, `TYP_SIMD12`, `TYP_SIMD16` and (where supported by the target) `TYP_SIMD32`.
+ - These types already exist, and represent some already-completed steps toward First Class Structs.
+
+ We propose to create the following additional types to be used where struct types of the given size are
+ passed and/or returned in registers:
+ * `TYP_STRUCT1`, `TYP_STRUCT2`, `TYP_STRUCT4`, `TYP_STRUCT8` (on 64-bit systems)
```

Comment:
I'm not sure that adding more struct types is a good thing. If anything, there are places in the JIT that have tables indexed by var_types and those tables will grow bigger. In particular, VN has AFAIR a bidimensional array where one of the dimensions is var_types and it's largish enough that adding/remove a type registers in the memory allocation stats.

I'm also not sure how these would interact with `ClassLayout` - would we need layout for a node having `TYP_STRUCT2`? I'd say yes because discarding layout/handle information has occassionally proved to be problematic. But then we end up with duplicate info - the size is encoded in both node's type and layout.

Reply:
Struct with specific alignment requirements - I guess we can simply add alignment info to `ClassLayout` :). Well, adding more info to it would make it bigger but the typical number of layouts a method needs is usually very low. And even if it's not it's still likely to be better than querying the VM multiple times or storing the same info in multiple places (e.g. in `fgArgTabEntry` objects).

Reply:
I've checked the memory usage issue I remembered about - it's there. Simply adding these 4 types to the list results in a 0.5% regression in memory usage, mostly thanks to VN. Perhaps there's a way to avoid that but...

At least on Linux x64, more structs can be returned in registers than on Windows x64 - a 3 byte struct is returned in `eax` and passed in, say, `edi`.

More generally, IMO it should be possible to enregister any structure small enough to fit in a register. You never known when it could be useful (think that for example on ARM you have `ubfx`/`sbfx` instructions which makes accessing fields of an enregistered structs cheap compared to x64 where you typically need 2 instructions to shift and zero/sign extended).

As such I don't think specific size `TYP_STRUCT` types help. It's not going to be very practical to later add even more `TYP_STRUCT3`, `TYP_STRUCT5` and so on.

I'll have to take a closer look at LSRA code but my expectancy is that for a `OBJ` node of suitable size the ref position/interval building code simply builds a ref position that asks for a suitable GPR (or why not, XMM) register and that's all there is to it.

I noticed that LSRA's interval has a `RegisterType registerType` where `RegisterType` is actually `var_types`. I find that a bit strange considering that ultimately there are just GPR and XMM registers and the finer distinction between types (`TYP_SHORT` vs. `TYP_USHORT`) is irrelevant to LSRA. Even requesting a byte sized registers on x86 isn't done by using `TYP_BYTE` but rather by restricting the register candidate set to those that support byte access.

And the funny thing is that LSRA too contributes to the memory regression that occurs when adding more types. It too has arrays indexed by type (e.g. `maxSpill`) even though only a couple of entries are actually used.